spec/src/lib.rs

```diff
@@ -81,18 +169,20 @@ pub(self) fn build_system_cell_transaction(
 }
 
 impl ChainSpec {
-    pub fn read_from_file<P: AsRef<Path>>(path: P) -> Result<ChainSpec, Box<Error>> {
-        let config_str = std::fs::read_to_string(path.as_ref())?;
+    pub fn read_from_file(spec_path: &SpecPath) -> Result<ChainSpec, Box<Error>> {
+        let config_bytes = spec_path.load_file(spec_path.path())?;
+        let config_str = String::from_utf8(config_bytes)?;
         let mut spec: Self = toml::from_str(&config_str)?;
-        spec.resolve_paths(path.as_ref().parent().expect("chain spec path resolve"));
+        spec.resolve_paths(spec_path.path().parent().expect("chain spec path resolve"));
+
         Ok(spec)
     }
 
     pub fn pow_engine(&self) -> Arc<dyn PowEngine> {
         self.pow.engine()
     }
 
-    pub fn to_consensus(&self) -> Result<Consensus, Box<Error>> {
+    pub fn to_consensus(&self, spec_path: &SpecPath) -> Result<Consensus, Box<Error>> {
```

Comment:
Refactoring to the builder pattern may simplify the code of `Consensus` creation.

Reply:
I actually tried that path, but we are talking about 2 parameters which are both required to build `Consensus`, I'm not sure using a builder pattern would simplify this code. It might end up just with more code and no obvious gain.

Tho I can refactor this to a separate function instead of method on ChainSpec.

Reply:
OK, I can try it later (someday... :)